Source/WebCore/Modules/url-pattern/URLPattern.cpp

```diff
@@ -26,29 +26,247 @@
+            String baseURLPath = processBaseURLString(baseURL.path(), type);
+            size_t slashIndex = baseURLPath.reverseFind('/');
+
+            if (slashIndex != notFound)
+                result.pathname = makeString(StringView { baseURLPath }.left(slashIndex + 1), result.pathname);
+
+            auto pathResult = canonicalizePath(result.pathname, baseURL.protocol(), type);
+
+            if (pathResult.hasException())
+                return pathResult.releaseException();
+
+            result.pathname = pathResult.releaseReturnValue();
+        }
+    }
+
+    if (!init.search.isNull()) {
+        auto queryResult = canonicalizeSearch(init.search, type);
+
+        if (queryResult.hasException())
+            return queryResult.releaseException();
+
+        result.search = queryResult.releaseReturnValue();
+    }
+
+    if (!init.hash.isNull()) {
+        auto fragmentResult = canonicalizeHash(init.hash, type);
+
+        if (fragmentResult.hasException())
+            return fragmentResult.releaseException();
+
+        result.hash = fragmentResult.releaseReturnValue();
+    }
+
+    return result;
+}
+
+ExceptionOr<Ref<URLPattern>> URLPattern::create(URLPatternInput&&, String&& baseURL, URLPatternOptions&&)
 {
     UNUSED_PARAM(baseURL);
 
+    // FIXME: Implement URLPattern::create
+
     return adoptRef(*new URLPattern());
 }
 
-Ref<URLPattern> URLPattern::create(std::optional<URLPatternInput>&&, URLPatternOptions&&)
+ExceptionOr<Ref<URLPattern>> URLPattern::create(std::optional<URLPatternInput>&& input, URLPatternOptions&&)
 {
-    return adoptRef(*new URLPattern());
+    URLPatternInit init;
+
+    if (std::holds_alternative<String>(input.value()) && !std::get<String>(input.value()).isNull()) {
+        // FIXME: Implement converting String input to URLPatternInit.
+    } else if (std::holds_alternative<URLPatternInit>(input.value()))
+        init = std::get<URLPatternInit>(input.value());
+
+    auto maybeProcessedInit = processInit(WTFMove(init), BaseURLStringType::Pattern, String(), String(), String(), String(), String(), String(), String(), String());
+
+    if (maybeProcessedInit.hasException())
+        return maybeProcessedInit.releaseException();
+
+    auto processedInit = maybeProcessedInit.releaseReturnValue();
+
+    // FIXME: Implement compiling the component for each URLPattern member field to replace this current placeholder code which reroutes processed URLPatternInit object to URLPattern result.
+    return adoptRef(*new URLPattern(WTFMove(processedInit.protocol), WTFMove(processedInit.username), WTFMove(processedInit.password), WTFMove(processedInit.hostname), WTFMove(processedInit.port), WTFMove(processedInit.pathname), WTFMove(processedInit.search), WTFMove(processedInit.hash)));
 }
 
-URLPattern::URLPattern() = default;
+URLPattern::URLPattern(String&& protocol, String&& username, String&& password, String&& hostname, String&& port, String&& pathname, String&& search, String&& hash)
```

Comment:
This constructor should likely just take a `URLPatternInit&&` to avoid so many parameters.